File: lang/nodejs/src/fun/vac/algorithms/search/BinarySearch.js
"use strict"

exports.find = (arr, key) => {
    let lo = 0
    let hi = arr.length

    let mid = 0

    while (lo < hi) {
        mid = lo + ((hi - lo) >>> 1)
        if (key < arr[mid]) {
            hi = mid
        } else if (key > arr[mid]) {
            lo = mid + 1
        } else {
            return mid
        }
    }

    return arr.length
}

File: go/muse/algorithms/search/BinarySearch/BinarySearch.go
package BinarySearch

func Find(arr []int, key int) int {
	lo := 0
	hi := len(arr)

	var mid int

	for lo < hi {
		mid = lo + ((hi - lo) >> 1)
		switch {
		case key < arr[mid]:
			hi = mid
		case key > arr[mid]:
			lo = mid + 1
		default:
			return mid
		}
	}

	return len(arr)
}
